### DISCUSSION TITLE
What is the significance of literal("a") vs "b" in concat example?

### DISCUSSION THREAD

**Discussion by anentropic** [sqlalchemy/sqlalchemy#9874] [Usage Questions] [answered]
https://docs.sqlalchemy.org/en/20/core/functions.html#sqlalchemy.sql.functions.concat

> String concatenation in SQLAlchemy is more commonly available using the Python + operator with string datatypes, which will render a backend-specific concatenation operator, such as :
```python
>>> print(select(literal("a") + "b"))
SELECT :param_1 || :param_2 AS anon_1
```

It's unclear to me in this example if:
- wrapping `"a"` in `literal(...)` is optional (as demonstrated by `"b"` doesn't have it)
- the first concat item has to be wrapped in `literal(...)`, perhaps for operator overloading reasons, but it's optional for subsequent items
- only the first item _must_ be wrapped in `literal(...)` and subsequent items _must not_ have it
- there's something contextual about why `literal("a")` but not `literal("b")` that I am not getting

**Comment by anentropic** [comment]
or... I guess `"b"` is an SQL column name, as python string, while `literal("a")` is the SQL string `"a"`

**Reply by anentropic** [reply]
> or... I guess `"b"` is an SQL column name, as python string, while `literal("a")` is the SQL string `"a"`

🤔  that idea made sense, but in fact they both seem to be treated as SQL string literals

```python
select(
    literal("local-") + "gamesession.id",
    ...
)
```

gives `[('local-gamesession.id', ...)]`

and

```python
select(
    literal("local-") + GameSession.id,
    ...
)
```

gives `[(3, ...)]` so the literal gets lost in this case


```python
select(
    text("'local-' || gamesession.id"),
    ...
)
```

works, but it's just raw SQL and not using SQLAlchemy concat feature

**Comment by CaselIT** [comment, marked as answer]
Hi,

literal is generated automatically when using a literal with any sqlalchemy construct. So `column('foo') + 1` is equivalent to `column('foo') + literal(1)`

In the example literal is explicit because otherwise the operation would be done by python: `select('a' + 'b')` is the same as `select('ab')`, meaning that it will generate `SELECT :param_1 AS anon_1`
